go-fuzz-dep/main.go

```diff
@@ -21,97 +17,24 @@ import (
 type Bool = bool
 
 var (
-	inFD  FD
-	outFD FD
-
 	CoverTab    *[CoverSize]byte
-	input       []byte
 	sonarRegion []byte
 	sonarPos    uint32
+	CoverTabTmp [CoverSize]byte
 )
 
-func init() {
-	var mem []byte
-	mem, inFD, outFD = setupCommFile()
-	CoverTab = (*[CoverSize]byte)(unsafe.Pointer(&mem[0]))
-	input = mem[CoverSize : CoverSize+MaxInputSize]
-	sonarRegion = mem[CoverSize+MaxInputSize:]
-}
 
-func Main(f func([]byte) int) {
-	runtime.GOMAXPROCS(1) // makes coverage more deterministic, we parallelize on higher level
-	for {
-		n := read(inFD)
-		if n > uint64(len(input)) {
-			println("invalid input length")
-			syscall.Exit(1)
-		}
-		for i := range CoverTab {
-			CoverTab[i] = 0
-		}
-		atomic.StoreUint32(&sonarPos, 0)
-		t0 := time.Now()
-		res := f(input[:n])
-		ns := time.Since(t0)
-		write(outFD, uint64(res), uint64(ns), uint64(atomic.LoadUint32(&sonarPos)))
-	}
-}
 
-// read reads little-endian-encoded uint64 from fd.
-func read(fd FD) uint64 {
-	rd := 0
-	var buf [8]byte
-	for rd != len(buf) {
-		n, err := fd.read(buf[rd:])
-		if err == syscall.EINTR {
-			continue
-		}
-		if n == 0 {
-			syscall.Exit(1)
-		}
-		if err != nil {
-			println("failed to read fd =", fd, "errno =", err.(syscall.Errno))
-			syscall.Exit(1)
-		}
-		rd += n
-	}
-	return deserialize64(buf[:])
+func init() {
+	CoverTab = (*[CoverSize]byte)(unsafe.Pointer(&CoverTabTmp[0]))
 }
 
-// write writes little-endian-encoded vals... to fd.
-func write(fd FD, vals ...uint64) {
-	var tmp [3 * 8]byte
-	buf := tmp[:len(vals)*8]
-	for i, v := range vals {
-		serialize64(buf[i*8:], v)
-	}
-	wr := 0
-	for wr != len(buf) {
-		n, err := fd.write(buf[wr:])
-		if err == syscall.EINTR {
-			continue
-		}
-		if err != nil {
-			println("failed to read fd =", fd, "errno =", err.(syscall.Errno))
-			syscall.Exit(1)
-		}
-		wr += n
+func Initialize(coverTabPtr unsafe.Pointer, coverTabSize uint64) {
```

Comment:
Could we just add this function to the package and leave everything else intact?

Reply:
I'd like to, but the ```init``` function is executed as soon as the binary is executed, and fails

```
failed to mmap fd = 3 errno = 9
```

Reply:
Yes, we should skip that init in libfuzzer mode.
Perhaps we could put code for normal mode and libfuzzer into separate files and use libfuzzer tag to select one or another in go-fuzz-build.